tags/views.py

```diff
@@ -505,14 +571,101 @@ def highlight(request):
   }
 
 '''
-Get all highlights for a page
+Add a highlight to a page
 '''
+@csrf_exempt
 @login_required
 @ajax_request
-def highlights(request):
+def highlight(request):
+  success = False
+  user = request.user
+  errors = {}
+  data = {}
+
+  if request.POST:
+    url = process_url(request.POST.get('url'))
+    highlight = request.POST.get('highlight')
+    highlight_id = request.POST.get('highlight_id')
+    errors['add_highlight'] = []
+
+    if highlight != "" or highlight_id:
+      p = Page.objects.get(url=url)
+
+      if highlight_id: 
+        try:
+          h = Highlight.objects.get(id=highlight_id)
+          data['highlight_id'] = h.id
+        except: 
+          errors['add_highlight'].append('Get highlight failed')
+      else:
+        try:
+          h, created = Highlight.objects.get_or_create(page=p, highlight=highlight)
+          h.user = user
+          h.save()
+
+          success = True
+          data['highlight_id'] = h.id
+        except:
+          errors['add_highlight'].append('Add highlight failed')
+
+  return {
+    'success': success,
+    'errors': errors,
+    'data': data,
+  }
+
+'''
+Get highlight by highlight id
+'''
+@login_required
+@ajax_request
+def highlight_by_id(request):
   success = False
   errors = {}
+  tags = {}
+  sorted_tags = []
   user = request.user
+  if user == None:
+    print "user not found"
+    return{
+    'success': False,
+    'errors': "user not found"
+    }
+  highlight = ''
+  highlight_owner = ''
+
+  if request.GET:
+    highlight_id = request.GET.get('highlight_id')
+    url = process_url(request.GET.get('url'))
+    errors['get_highlights'] = []
+
+    if not len(errors['get_highlights']):
+      highlight = Highlight.objects.get(id=highlight_id).highlight
+      highlight_owner =  Highlight.objects.get(id=highlight_id).user.username
+      success = True
+
+  return {
+    'success': success,
+    'errors': errors,
+    'highlight': highlight,
+    'highlight_owner':highlight_owner,
+  }
+
+'''
+Get all highlights for a page using value tags
+'''
+@login_required
+@ajax_request
+def highlights_tags(request):
```

Comment:
ditto for this. where is the endpoint in urls.py? Where do you actually call this endpoint?

Reply:
also the different but very similar sounding endpoints of /highlight, /highlights, /highlight_tags, and /highlights_tags are pretty confusing. can you rename and add comments about what each does?